### ISSUE TITLE
export: allow users to define gene order in genotype color-by 

### ISSUE THREAD

**Issue by joverlee521** [nextstrain/augur#1052]
### Context
Related to https://github.com/nextstrain/auspice/issues/1549

### Description
`augur export v2` should allow users to define the gene order of the genotype color-by through the auspice-config.json

**Comment by jameshadfield** [comment]
I looked into this briefly today, as I had always assumed these were ordered according to their position in the genome. 

`augur translate` reads in annotations and preserves [1] the order they appear in the genbank file (didn't test VCF). This order is typically the order they appear in the genome. When these are written out they are sorted (alphabetically) via: 

https://github.com/nextstrain/augur/blob/e1e8bb155c25a2901bcd41423b0f67d14386b6b3/augur/utils.py#L125-L127

so if we use an OrderedDict instead for the translate output, i.e. change

https://github.com/nextstrain/augur/blob/e1e8bb155c25a2901bcd41423b0f67d14386b6b3/augur/translate.py#L414


to `output_data = OrderedDict({'annotations':annotations, 'nodes':aa_muts})` then the `aa_muts.json` file will preserve the gene order of the GenBank file. `augur export v2` will preserve the order of keys as it uses `OrderedDict`s internally. 


I think we should use an OrderedDict within `augur translate` which will lead to the auspice dropdown using the gene ordering in the GenBank file. This is normally the order they appear in the genome which is a more sensible default than an alphabetical sort. It also allows us to simply change the order of genes within the genbank file if we want to change this ordering. I prefer this to adding more complexity to the auspice-config JSON (although I don't feel particularly strongly) which is in line with my idea to encode more information in the node-data JSONs themselves (see https://github.com/nextstrain/augur/issues/988 for more). 

[1] Guaranteed from python 3.6, I think, but was the typical behavior before then.

**Comment by emmahodcroft** [comment]
I think this approach is very sensible - alphabetical doesn't really make much sense. My only question would be whether it's worth it to have a 'shortcut' way to specify order, if we want to just do something like move Spike up front for SC2. I don't know how much we modify the Genbank reference at the moment - maybe it's good to have this match what one would download from Genbank to avoid confusion?

If we already modify it though, I think it's fine to then say we just change the gene order there. If not, it might be nice to have a way to specify this order somewhere (either in Auspice or when translating?) so that one can keep the reference untouched if that's been our stance so far?

**Comment by jameshadfield** [comment]
Yes, good points. I can imagine modifying canonical GenBank files may not be appropriate, especially as my guess is a lot of people are using the reference from our ncov pipeline for their own work. 

>  If we want to just do something like move Spike up front for SC2

The following (on top of the 1-line change to augur above) will be enough for one-off cases like nCoV:

```py
rule change_gene_order:
    input: rules.translate.output.node_data,
    output: "results/aa_muts_ordered.json",
    run:
        import json
        with open(input[0]) as fh:
            data = json.load(fh)
        order = ["S"]
        data['annotations'] = {k:data['annotations'][k] for k in order+[x for x in data['annotations'].keys() if x not in order]}
        with open(output[0], 'w', encoding='utf-8') as fh:
            json.dump(data, fh, indent=2)
```

Or put it in a little script and then pipe the output of `augur translate` through it, so it's all part of `rule.translate` is perhaps cleaner than adding (yet) another rule.

**Comment by jameshadfield** [comment]
Note to self: revisit `areNucleotideMutationsPresent` function in auspice (added by auspice PR linked above) if we move to OrderedDict here.

**Comment by emmahodcroft** [comment]
>so that it can be easily switched and the only command that would need to rerun is augur export.

I can definitely see the benefit of that!

**Comment by jameshadfield** [comment]
> I would prefer all of the auspice config to be in a place that is editable outside of the pipeline so that it can be easily switched and the only command that would need to rerun is augur export

This is in line with our current framework, so I'm happy to go for it. I don't know if it's the _best_, but it's consistent with what we do so let's stay on this path (I haven't fully thought out the alternative, either).
